### PR TITLE
HIR-123. Add a booking service.

### PR DESCRIPTION
**Pull request by oleg1std**
Add a booking controller.
Add a booking service.

### CODE REVIEW

File: HappyTravel.Hiroshima.WebApi/Services/AvailabilitySearch/BookingResponseService.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public class BookingResponseService : IBookingResponseService
+    {
+        public EdoContracts.Accommodations.Booking Create(Common.Models.Bookings.BookingOrder bookingOrder)
+        {
+            var referenceCode = bookingOrder.ReferenceCode;
+            var status = GetStatus(bookingOrder.Status);
+            var availabilityRequest = bookingOrder.AvailabilityRequest.GetValue<EdoContracts.Accommodations.AvailabilityRequest>();
+            var availableRates = bookingOrder.AvailableRates.GetValue<Common.Models.Availabilities.AvailableRatesSlim>();
+            var bookingRequest = bookingOrder.BookingRequest .GetValue<EdoContracts.Accommodations.BookingRequest>();
```
Comment (kirillta) [inline HappyTravel.Hiroshima.WebApi/Services/AvailabilitySearch/BookingResponseService.cs:15]: Unnecessary space

File: HappyTravel.Hiroshima.Common/Models/Accommodations/Rooms/OccupancyDefinitions/OccupancyConfiguration.cs [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 namespace HappyTravel.Hiroshima.Common.Models.Accommodations.Rooms.OccupancyDefinitions
 {
+    [Serializable]
```
Comment (kirillta) [inline HappyTravel.Hiroshima.Common/Models/Accommodations/Rooms/OccupancyDefinitions/OccupancyConfiguration.cs:5]: Do we need the attribute?

File: HappyTravel.Hiroshima.Common/Models/ContactInfo.cs [hunk 3]
```diff
@@ -7,5 +9,14 @@ public class ContactInfo
         public string Phone { get; set; } = string.Empty;
 
         public string Website { get; set; } = string.Empty;
```
Comment (kirillta) [inline HappyTravel.Hiroshima.Common/Models/ContactInfo.cs:11]: Could we change contact's property types to lists to fit the contracts?

File: HappyTravel.Hiroshima.DirectContracts/Extensions/FunctionalExensions/ResultTransactionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
+﻿using System;
+using System.Threading.Tasks;
+using CSharpFunctionalExtensions;
+using HappyTravel.Hiroshima.Data;
+
+namespace HappyTravel.Hiroshima.Common.Infrastructure.Extensions.Extensions.FunctionalExensions
+{
+    public static class ResultTransactionExtensions
+    {
+        public static Task<Result<T>> BindWithTransaction<T>(
```
Comment (kirillta) [inline HappyTravel.Hiroshima.DirectContracts/Extensions/FunctionalExensions/ResultTransactionExtensions.cs:10]: Maybe we need to move things like that to a package @oleg1std @mingazhev

File: HappyTravel.Hiroshima.DirectContracts/Services/BookingService.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            _availabilityDataStorage = availabilityDataStorage;
+            _availabilityService = availabilityService;
+            _dbContext = dbContext;
+        }
+
+
+        public async Task<Result<Common.Models.Bookings.BookingOrder>> Book(EdoContracts.Accommodations.BookingRequest bookingRequest, EdoContracts.Accommodations.AvailabilityRequest availabilityRequest, string languageCode)
+        {
+            const string error = "There are no available rates";
```
Comment (kirillta) [inline HappyTravel.Hiroshima.DirectContracts/Services/BookingService.cs:28]: This isn't clear. No rates for like what?

File: HappyTravel.Hiroshima.WebApi/Controllers/Connector/BookingController.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            var (_, isFailure, bookingDetails, error) = await _bookingService.Book(bookingRequest, LanguageCode);
+            if (isFailure)
+                return BadRequest(ProblemDetailsBuilder.Build(error));
+
+            return Ok(bookingDetails);
+        }
+
+
+        private readonly IBookingService _bookingService;
+        private string LanguageCode => CultureInfo.CurrentCulture.Name;
```
Comment (kirillta) [inline HappyTravel.Hiroshima.WebApi/Controllers/Connector/BookingController.cs:44]: This field could be moved to a base class

File: HappyTravel.Hiroshima.DirectContracts/Extensions/FunctionalExensions/ResultTransactionExtensions.cs (hunk 4 above)
Comment (oleg1std) [reply HappyTravel.Hiroshima.DirectContracts/Extensions/FunctionalExensions/ResultTransactionExtensions.cs:10]: Looks reasonable in general. Maybe.

**Review by AlexeyShok — APPROVED** [review]

File: HappyTravel.Hiroshima.Common/Models/ContactInfo.cs [hunk 7]
```diff
@@ -1,11 +1,26 @@
...
+using System.Collections.Generic;
+using HappyTravel.EdoContracts.Extensions;
+using HappyTravel.Hiroshima.Common.Infrastructure.Utilities;
+
+namespace HappyTravel.Hiroshima.Common.Models
 {
     public class ContactInfo
     {
-        public string Email { get; set; } = string.Empty;
+        public List<string> Email { get; set; } = new List<string>();
```
Comment (kirillta) [inline HappyTravel.Hiroshima.Common/Models/ContactInfo.cs:10]: Emails, phones, websites are in plural of course

**Review by kirillta — APPROVED** [review]

**Pull request merged**